### PR TITLE
added image patch grad

### PR DESCRIPTION
**Pull request by nikhilmishra000**
This solves: https://github.com/tensorflow/tensorflow/issues/2921

I ran unit tests using tensorflow.python.ops.gradient_checker to verify that it works :D

### CODE REVIEW

File: tensorflow/python/ops/array_grad.py [hunk 1]
```diff
@@ -448,3 +451,77 @@ def _MirrorPadGradGrad(op, grad):
 @ops.RegisterGradient("QuantizeAndDequantize")
 def _QuantizeAndDequantizeGrad(_, grad):
   return grad
+
+
+@ops.RegisterGradient("ExtractImagePatches")
+def _ExtractImagePatchesGrad(op, grad):
```
Comment (benoitsteiner) [inline tensorflow/python/ops/array_grad.py]: Thanks for the contribution, this looks pretty good. Can you add a regression test to make sure the code is correct? You can use https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/conv2d_backprop_filter_grad_test.py#L25 as an example.

**Comment by nikhilmishra000** [comment]
added the tests, let me know if there's anything else

**Comment by danmane** [comment]
@tensorflow-jenkins test this please

**Comment by benoitsteiner** [comment]
Jenkins, test this please.

**Comment by danmane** [comment]
The tests are failing :(

**Comment by nikhilmishra000** [comment]
I'm looking at the logs, and the one that's failing is //tensorflow/tools/test:check_futures_test.
I think it's because my test file "tensorflow/python/kernel_tests/extract_image_patches_grad_test.py" doesn't import "absolute_import"?

**Comment by danmane** [comment]
Push the fix, and I'll re-trigger the tests.

**Comment by danmane** [comment]
@tensorflow-jenkins test this please

**Comment by danmane** [comment]
@benoitsteiner Can you LGTM this now that tests were added?

File: tensorflow/python/ops/array_grad.py [hunk 2]
```diff
@@ -448,3 +451,78 @@ def _MirrorPadGradGrad(op, grad):
...
+  ksize_r_eff = ksize_r + (ksize_r - 1) * (rate_r - 1)
+  ksize_c_eff = ksize_c + (ksize_c - 1) * (rate_c - 1)
+
+  if padding == 'SAME':
+    rows_out = int(ceil(rows_in / stride_r))
+    cols_out = int(ceil(cols_in / stride_h))
+    pad_rows = ((rows_out - 1) * stride_r + ksize_r_eff - rows_in) // 2
+    pad_cols = ((cols_out - 1) * stride_h + ksize_c_eff - cols_in) // 2
+
+  elif padding == 'VALID':
```
Comment (gunan) [inline tensorflow/python/ops/array_grad.py]: Looks like in python 3 PIP setups, there is a problem caused by these lines.
It looks like "padding" can be something other than 'SAME' or 'VALID'
any ideas what might be happening?

http://ci.tensorflow.org/view/Tensorflow%20Jenkins%20Monitored%20builds/job/nightly-matrix-cpu/TF_BUILD_CONTAINER_TYPE=CPU,TF_BUILD_IS_OPT=OPT,TF_BUILD_IS_PIP=PIP,TF_BUILD_PYTHON_VERSION=PYTHON3,label=cpu-slave/194/consoleFull

**Comment by gunan** [comment]
Looks like the BUILD file change for the test is missing from this PR.
Therefore Jenkins did not run the new feature, and I suspect this is broken.
I am investigating now.

**Comment by gunan** [comment]
I am preparing a fix for the issue.
It will be pushed to github later today, or tomorrow.

**Comment by mohamedadaly** [comment]
The current gradient seems to have a problem when `batches` is `None` i.e. when using a variable size batches using Keras. Simply added this line
```python
  batch_size = array_ops.shape(op.inputs[0])[0]
```
after this
```python
  batch_size, rows_in, cols_in, channels = [
    dim.value for dim in op.inputs[0].get_shape()
  ]
```
and it seems to have fixed the problem. I am new to TensorFlow, so you probably have a better solution than this.
